fetch/fetch.go

```diff
@@ -300,6 +300,12 @@ func NewFetch(
 	if host != nil {
 		connectedf := func(peer p2p.Peer) {
 			protocols := func() []protocol.ID {
+				// Make sure that the protocol list for the peer is correct.
+				// This is similar to what Host.NewStream does to make
+				// sure it is possible to use one of the specified
+				// protocols. If we don't do this, there may be a race causing
+				// some peers to be unnecessarily ignored.
+				host.Identify(peer)
```

Comment:
I believe this:

```suggestion
				pi := host.Peerstore().PeerInfo(peer)
				if err := host.Connect(context.Background(), pi); err != nil {
					f.logger.Debug("failed to connect to peer",
						zap.Stringer("id", peer),
						zap.Error(err),
					)
					return nil
				}
```

has basically the same effect without a) needing to intercept the building process of the `libp2p` host and b) allowing us to pass a timeout to `Connect` in case we want to abort early if something goes wrong.

`Connect` in both implementations of `Host` eventually calls `IdentifyWait` which is also called by `IdentifyCon` but allows passing a context in case we want to abort early.

Wdyt?

Reply:
This will not work b/c `Connect` is noop in case if the peer is already connected:
https://github.com/libp2p/go-libp2p/blob/v0.38.1/p2p/host/basic/basic_host.go#L803
So it will not call `IdentifyWait` in this case (it is invoked from `dialPeer`: https://github.com/libp2p/go-libp2p/blob/v0.38.1/p2p/host/basic/basic_host.go#L826)
And if you force dial via a context option, there will be adverse side effects such as trying to establish a new connection, invoking the gater etc.

Reply:
There may be a way of trying to create a new stream which also calls `IdentifyWait`, but I do not like it either as it adds overhead of creating a new stream when it's not really needed, producing unwanted network traffic

Reply:
I'm not sure I understand. For the network status to be connected doesn't this mean `Connect` has already been called and it is OK that `Connect` is a no-op? The first call to it result in `IdentifyWait` being called?

Again to me this feels wrong - pulling internals out of libp2p that look like they might not be intended to be used outside of the library (`IdentifyConn` looks like a helper method for tests of the `IDService`)

Reply:
The peer is listed in the addressbook before `Connect` finishes.
The need to access internals is rather unfortunate but it's due to some inconvenient decisions in `go-libp2p` codebase. Basically `IDService` can be easily accessed if you know for sure your `Host` is a `*BasicHost`, but it is harder to reach if there are some `Host` wrappers. So `IDService` is not fully internal.
`IdentifyConn` just calls `IdentifyWait` (which is not only used in tests) and waits on the channel it returns

Reply:
If we have to use IDService because there are no alternatives I would still prefer to use `IdentifyWait` over `IdentifyConn` because then we can at least pass a timeout instead of (possibly) blocking indefinitely.